// @scaleton/highload-wallet/dist/BatchSender.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchSender = void 0;
const core_1 = require("@ton/core");
class BatchSender {
    constructor(provider, wallet, secretKey, batchSize) {
        this.provider = provider;
        this.wallet = wallet;
        this.secretKey = secretKey;
        this.batchSize = batchSize;
        this.messages = [];
    }
    get address() {
        return this.wallet.address;
    }
    clear() {
        this.messages.splice(0, this.messages.length);
    }
    async send(args) {
        this.messages.push([
            (0, core_1.internal)({
                to: args.to,
                value: args.value,
                init: args.init,
                body: args.body,
                bounce: args.bounce,
            }),
            args.sendMode ?? core_1.SendMode.PAY_GAS_SEPARATELY,
        ]);
    }
    async submit() {
        for (let i = 0; i < this.messages.length; i += this.batchSize) {
            await this.wallet.sendTransfer(this.provider, {
                queryId: this.wallet.generateQueryId(),
                messages: this.messages.slice(i, i + this.batchSize),
                secretKey: this.secretKey,
            });
        }
        this.clear();
    }
}
exports.BatchSender = BatchSender;
